File: backend/src/plugins/risk-checker.js
const config = require('../utils/config');
const logger = require('../utils/logger');
const User = require('../models/User');

class RiskCheckerPlugin {
  constructor() {
    this.name = 'risk-checker';
    this.version = '1.0.0';
    this.description = 'Risk assessment and fraud detection for transactions';
    
    // Risk thresholds and rules
    this.riskThresholds = {
      highValue: config.plugins.riskThresholdAmount || 10000,
      velocityLimit: 5, // Max transactions per hour
      velocityAmount: 20000, // Max amount per hour
      suspiciousCountries: ['XX', 'YY'], // Mock suspicious country codes
      offHours: { start: 23, end: 6 }, // 11 PM to 6 AM
      maxDailyAmount: 50000
    };

    // Risk scoring weights
    this.riskWeights = {
      amount: 0.3,
      velocity: 0.25,
      time: 0.15,
      location: 0.15,
      merchant: 0.1,
      user: 0.05
    };
  }

  /**
   * Check if plugin should be enabled for this transaction
   */
  isEnabled(transaction, context) {
    // Risk checking is enabled for all transactions except test transactions
    return context.skipRiskCheck !== true;
  }

  /**
   * Process risk assessment
   */
  async process(transaction, context) {
    const startTime = Date.now();
    
    try {
      logger.info(`Risk Checker processing transaction ${transaction.txnId}`, {
        amount: transaction.amount,
        currency: transaction.currency,
        type: transaction.type
      });

      // Perform various risk checks
      const riskAssessment = await this.performRiskAssessment(transaction, context);
      
      // Determine action based on risk score
      const action = this.determineAction(riskAssessment);
      
      // Update transaction status if blocked
      if (action.block) {
        await transaction.updateStatus('blocked', {
          riskScore: riskAssessment.totalScore,
          blockReason: action.reason,
          riskFactors: riskAssessment.factors
        });
      } else if (action.flag) {
        await transaction.updateStatus('flagged', {
          riskScore: riskAssessment.totalScore,
          flagReason: action.reason,
          riskFactors: riskAssessment.factors
        });
      }

      const result = {
        success: true,
        action: action.block ? 'blocked' : (action.flag ? 'flagged' : 'approved'),
        riskScore: riskAssessment.totalScore,
        riskLevel: this.getRiskLevel(riskAssessment.totalScore),
        factors: riskAssessment.factors,
        recommendation: action.reason,
        processingTime: Date.now() - startTime,
        requiresManualReview: action.flag || riskAssessment.totalScore > 70,
        transaction: action.block ? { status: 'blocked' } : null
      };

      logger.info(`Risk assessment completed for transaction ${transaction.txnId}`, {
        riskScore: riskAssessment.totalScore,
        action: result.action,
        factors: riskAssessment.factors.length
      });

      return result;

    } catch (error) {
      logger.logError(error, { 
        action: 'riskAssessment', 
        txnId: transaction.txnId 
      });

      return {
        success: false,
        action: 'failed',
        error: error.message,
        riskScore: 0,
        processingTime: Date.now() - startTime
      };
    }
  }

  /**
   * Perform comprehensive risk assessment
   */
  async performRiskAssessment(transaction, context) {
    const factors = [];
    let totalScore = 0;

    // 1. Amount-based risk
    const amountRisk = this.assessAmountRisk(transaction);
    if (amountRisk.score > 0) {
      factors.push(amountRisk);
      totalScore += amountRisk.score * this.riskWeights.amount;
    }

    // 2. Velocity risk (transaction frequency)
    const velocityRisk = await this.assessVelocityRisk(transaction);
    if (velocityRisk.score > 0) {
      factors.push(velocityRisk);
      totalScore += velocityRisk.score * this.riskWeights.velocity;
    }

    // 3. Time-based risk (off-hours transactions)
    const timeRisk = this.assessTimeRisk(transaction);
    if (timeRisk.score > 0) {
      factors.push(timeRisk);
      totalScore += timeRisk.score * this.riskWeights.time;
    }

    // 4. Location/Merchant risk
    const merchantRisk = this.assessMerchantRisk(transaction, context);
    if (merchantRisk.score > 0) {
      factors.push(merchantRisk);
      totalScore += merchantRisk.score * this.riskWeights.merchant;
    }

    // 5. User profile risk
    const userRisk = await this.assessUserRisk(transaction);
    if (userRisk.score > 0) {
      factors.push(userRisk);
      totalScore += userRisk.score * this.riskWeights.user;
    }

    return {
      totalScore: Math.min(Math.round(totalScore), 100), // Cap at 100
      factors,
      assessmentTime: new Date().toISOString()
    };
  }

  /**
   * Assess risk based on transaction amount
   */
  assessAmountRisk(transaction) {
    const amount = parseFloat(transaction.convertedAmount || transaction.amount);
    let score = 0;
    let reason = '';

    if (amount > this.riskThresholds.highValue) {
      score = Math.min(50 + (amount - this.riskThresholds.highValue) / 1000, 100);
      reason = `High value transaction: ${transaction.currency} ${amount}`;
    } else if (amount > this.riskThresholds.highValue * 0.5) {
      score = 25;
      reason = `Medium value transaction: ${transaction.currency} ${amount}`;
    }

    return score > 0 ? {
      type: 'amount',
      score: Math.round(score),
      reason,
      details: { amount, currency: transaction.currency, threshold: this.riskThresholds.highValue }
    } : { score: 0 };
  }

  /**
   * Assess velocity risk (transaction frequency)
   */
  async assessVelocityRisk(transaction) {
    try {
      // Get user's recent transactions (last hour)
      const oneHourAgo = new Date(Date.now() - 60 * 60 * 1000);
      
      // In a real implementation, you'd query the database for recent transactions
      // For demo purposes, we'll simulate this
      const recentTransactions = await this.getRecentTransactions(transaction.userId, oneHourAgo);
      
      let score = 0;
      let reason = '';

      if (recentTransactions.count > this.riskThresholds.velocityLimit) {
        score = Math.min(40 + (recentTransactions.count - this.riskThresholds.velocityLimit) * 10, 100);
        reason = `High transaction velocity: ${recentTransactions.count} transactions in last hour`;
      }

      if (recentTransactions.totalAmount > this.riskThresholds.velocityAmount) {
        const velocityScore = Math.min(30 + (recentTransactions.totalAmount - this.riskThresholds.velocityAmount) / 1000, 100);
        if (velocityScore > score) {
          score = velocityScore;
          reason = `High velocity amount: ${recentTransactions.totalAmount} in last hour`;
        }
      }

      return score > 0 ? {
        type: 'velocity',
        score: Math.round(score),
        reason,
        details: { 
          transactionCount: recentTransactions.count, 
          totalAmount: recentTransactions.totalAmount,
          timeWindow: '1 hour'
        }
      } : { score: 0 };

    } catch (error) {
      logger.logError(error, { action: 'assessVelocityRisk', userId: transaction.userId });
      return { score: 0 };
    }
  }

  /**
   * Assess time-based risk
   */
  assessTimeRisk(transaction) {
    const now = new Date();
    const hour = now.getHours();
    
    let score = 0;
    let reason = '';

    // Check if transaction is during off-hours
    if (hour >= this.riskThresholds.offHours.start || hour <= this.riskThresholds.offHours.end) {
      score = 20;
      reason = `Off-hours transaction at ${hour}:${now.getMinutes().toString().padStart(2, '0')}`;
    }

    // Weekend transactions (additional risk)
    const dayOfWeek = now.getDay();
    if (dayOfWeek === 0 || dayOfWeek === 6) { // Sunday or Saturday
      score += 10;
      reason += (reason ? '; ' : '') + 'Weekend transaction';
    }

    return score > 0 ? {
      type: 'time',
      score,
      reason,
      details: { hour, dayOfWeek, timestamp: now.toISOString() }
    } : { score: 0 };
  }

  /**
   * Assess merchant/location risk
   */
  assessMerchantRisk(transaction, context) {
    let score = 0;
    let reason = '';

    // Check for suspicious merchant patterns
    if (transaction.merchantId) {
      // Mock merchant risk assessment
      if (transaction.merchantId.includes('suspicious') || transaction.merchantId.includes('test')) {
        score = 30;
        reason = 'Potentially suspicious merchant';
      }
    }

    // Check for suspicious location (from context)
    if (context.userLocation && this.riskThresholds.suspiciousCountries.includes(context.userLocation.country)) {
      score = Math.max(score, 40);
      reason = `Transaction from high-risk location: ${context.userLocation.country}`;
    }

    return score > 0 ? {
      type: 'merchant',
      score,
      reason,
      details: { 
        merchantId: transaction.merchantId, 
        merchantName: transaction.merchantName,
        location: context.userLocation 
      }
    } : { score: 0 };
  }

  /**
   * Assess user profile risk
   */
  async assessUserRisk(transaction) {
    try {
      const user = await User.findById(transaction.userId);
      if (!user) {
        return {
          type: 'user',
          score: 50,
          reason: 'User not found',
          details: { userId: transaction.userId }
        };
      }

      let score = 0;
      let reason = '';

      // New user risk
      const userAge = Date.now() - new Date(user.createdAt).getTime();
      const daysSinceCreation = userAge / (1000 * 60 * 60 * 24);
      
      if (daysSinceCreation < 7) {
        score = 25;
        reason = `New user account (${Math.round(daysSinceCreation)} days old)`;
      }

      // Account status risk
      if (user.status !== 'active') {
        score = Math.max(score, 60);
        reason = `User account status: ${user.status}`;
      }

      return score > 0 ? {
        type: 'user',
        score,
        reason,
        details: { 
          userId: user.id, 
          accountAge: Math.round(daysSinceCreation),
          status: user.status 
        }
      } : { score: 0 };

    } catch (error) {
      logger.logError(error, { action: 'assessUserRisk', userId: transaction.userId });
      return {
        type: 'user',
        score: 30,
        reason: 'Unable to assess user risk',
        details: { error: error.message }
      };
    }
  }

  /**
   * Determine action based on risk assessment
   */
  determineAction(riskAssessment) {
    const score = riskAssessment.totalScore;

    if (score >= 80) {
      return {
        block: true,
        flag: false,
        reason: 'High risk transaction - automatically blocked'
      };
    } else if (score >= 60) {
      return {
        block: false,
        flag: true,
        reason: 'Medium-high risk transaction - flagged for manual review'
      };
    } else if (score >= 40) {
      return {
        block: false,
        flag: true,
        reason: 'Medium risk transaction - flagged for monitoring'
      };
    } else {
      return {
        block: false,
        flag: false,
        reason: 'Low risk transaction - approved'
      };
    }
  }

  /**
   * Get risk level description
   */
  getRiskLevel(score) {
    if (score >= 80) return 'CRITICAL';
    if (score >= 60) return 'HIGH';
    if (score >= 40) return 'MEDIUM';
    if (score >= 20) return 'LOW';
    return 'MINIMAL';
  }

  /**
   * Mock function to get recent transactions (would query database in production)
   */
  async getRecentTransactions(userId, since) {
    // Mock data - in production this would query the database
    return {
      count: Math.floor(Math.random() * 3), // 0-2 recent transactions
      totalAmount: Math.floor(Math.random() * 5000) // Random amount
    };
  }

  /**
   * Update risk thresholds (for admin configuration)
   */
  updateRiskThresholds(newThresholds) {
    this.riskThresholds = { ...this.riskThresholds, ...newThresholds };
    logger.info('Risk thresholds updated', newThresholds);
  }

  /**
   * Get current risk configuration
   */
  getRiskConfiguration() {
    return {
      thresholds: this.riskThresholds,
      weights: this.riskWeights,
      version: this.version
    };
  }

  /**
   * Check if this plugin is critical
   */
  isCritical(transaction, context) {
    // Risk checking is critical for high-value transactions
    return parseFloat(transaction.amount) > this.riskThresholds.highValue;
  }
}

module.exports = RiskCheckerPlugin; 

File: backend/src/plugins/risk_checker.py
"""
Risk Checker Plugin for SatuPay Payment Switch
Analyzes transaction risk and applies risk management rules
"""

import asyncio
from datetime import datetime, timedelta
from typing import Dict, Any, List
from .plugin_manager import BasePlugin

class RiskCheckerPlugin(BasePlugin):
    """Risk Checker Plugin for transaction risk assessment"""
    
    def __init__(self):
        super().__init__()
        self.name = "Risk Checker"
        self.version = "1.0.0"
        self.description = "Transaction risk assessment and fraud detection"
        self.critical = False
        
        # Risk thresholds
        self.risk_thresholds = {
            'high_amount': 10000.0,  # MYR
            'very_high_amount': 50000.0,  # MYR
            'suspicious_frequency': 10,  # transactions per hour
            'velocity_limit': 50000.0,  # MYR per day
            'international_limit': 25000.0,  # MYR for international
            'max_risk_score': 85  # Block if risk score > 85
        }
        
        # Risk factors weights
        self.risk_weights = {
            'amount_factor': 0.3,
            'frequency_factor': 0.25,
            'velocity_factor': 0.2,
            'time_factor': 0.1,
            'merchant_factor': 0.1,
            'location_factor': 0.05
        }
        
        self.config = {
            'enabled_checks': [
                'amount_check',
                'frequency_check', 
                'velocity_check',
                'time_pattern_check',
                'merchant_reputation_check'
            ],
            'auto_block_threshold': 85,
            'manual_review_threshold': 70,
            'risk_thresholds': self.risk_thresholds
        }
    
    async def execute(self, data: Dict[str, Any]) -> Dict[str, Any]:
        """Execute risk assessment"""
        try:
            risk_assessment = await self._perform_risk_assessment(data)
            
            # Determine action based on risk score
            action = self._determine_action(risk_assessment['risk_score'])
            
            return {
                'success': True,
                'data': {
                    'risk_assessment': risk_assessment,
                    'risk_action': action,
                    'risk_checks_completed': True
                }
            }
            
        except Exception as e:
            return {
                'success': False,
                'error': f"Risk assessment failed: {str(e)}",
                'critical': False
            }
    
    async def _perform_risk_assessment(self, data: Dict[str, Any]) -> Dict[str, Any]:
        """Perform comprehensive risk assessment"""
        
        risk_factors = {}
        total_risk_score = 0
        
        # Amount-based risk
        if 'amount_check' in self.config['enabled_checks']:
            amount_risk = await self._check_amount_risk(data)
            risk_factors['amount_risk'] = amount_risk
            total_risk_score += amount_risk['score'] * self.risk_weights['amount_factor']
        
        # Frequency-based risk
        if 'frequency_check' in self.config['enabled_checks']:
            frequency_risk = await self._check_frequency_risk(data)
            risk_factors['frequency_risk'] = frequency_risk
            total_risk_score += frequency_risk['score'] * self.risk_weights['frequency_factor']
        
        # Velocity-based risk
        if 'velocity_check' in self.config['enabled_checks']:
            velocity_risk = await self._check_velocity_risk(data)
            risk_factors['velocity_risk'] = velocity_risk
            total_risk_score += velocity_risk['score'] * self.risk_weights['velocity_factor']
        
        # Time pattern risk
        if 'time_pattern_check' in self.config['enabled_checks']:
            time_risk = await self._check_time_pattern_risk(data)
            risk_factors['time_risk'] = time_risk
            total_risk_score += time_risk['score'] * self.risk_weights['time_factor']
        
        # Merchant reputation risk
        if 'merchant_reputation_check' in self.config['enabled_checks']:
            merchant_risk = await self._check_merchant_reputation_risk(data)
            risk_factors['merchant_risk'] = merchant_risk
            total_risk_score += merchant_risk['score'] * self.risk_weights['merchant_factor']
        
        return {
            'risk_score': min(round(total_risk_score, 1), 100),
            'risk_level': self._get_risk_level(total_risk_score),
            'risk_factors': risk_factors,
            'assessment_timestamp': datetime.utcnow().isoformat()
        }
    
    async def _check_amount_risk(self, data: Dict[str, Any]) -> Dict[str, Any]:
        """Check amount-based risk"""
        amount = float(data.get('amount', 0))
        currency = data.get('currency', 'MYR')
        
        # Convert to MYR if needed (simplified)
        if currency != 'MYR':
            # Use rough conversion rates for risk assessment
            conversion_rates = {'USD': 4.7, 'SGD': 3.5, 'EUR': 5.1}
            amount = amount * conversion_rates.get(currency, 1)
        
        risk_score = 0
        risk_factors = []
        
        if amount >= self.risk_thresholds['very_high_amount']:
            risk_score = 80
            risk_factors.append(f"Very high amount: {amount:.2f} MYR")
        elif amount >= self.risk_thresholds['high_amount']:
            risk_score = 50
            risk_factors.append(f"High amount: {amount:.2f} MYR")
        elif amount >= self.risk_thresholds['high_amount'] * 0.5:
            risk_score = 25
            risk_factors.append(f"Moderate amount: {amount:.2f} MYR")
        else:
            risk_score = 5
            risk_factors.append(f"Normal amount: {amount:.2f} MYR")
        
        return {
            'score': risk_score,
            'factors': risk_factors,
            'amount_myr': amount
        }
    
    async def _check_frequency_risk(self, data: Dict[str, Any]) -> Dict[str, Any]:
        """Check frequency-based risk (mock implementation)"""
        # In production, this would query the database for recent transactions
        await asyncio.sleep(0.05)  # Simulate DB query
        
        # Mock frequency data
        user_id = data.get('user_id')
        recent_txn_count = 3  # Mock: user has 3 transactions in last hour
        
        risk_score = 0
        risk_factors = []
        
        if recent_txn_count >= self.risk_thresholds['suspicious_frequency']:
            risk_score = 70
            risk_factors.append(f"Suspicious frequency: {recent_txn_count} txns/hour")
        elif recent_txn_count >= self.risk_thresholds['suspicious_frequency'] * 0.6:
            risk_score = 40
            risk_factors.append(f"High frequency: {recent_txn_count} txns/hour")
        elif recent_txn_count >= 5:
            risk_score = 20
            risk_factors.append(f"Moderate frequency: {recent_txn_count} txns/hour")
        else:
            risk_score = 5
            risk_factors.append(f"Normal frequency: {recent_txn_count} txns/hour")
        
        return {
            'score': risk_score,
            'factors': risk_factors,
            'recent_transaction_count': recent_txn_count
        }
    
    async def _check_velocity_risk(self, data: Dict[str, Any]) -> Dict[str, Any]:
        """Check velocity-based risk (daily spending limit)"""
        # Mock implementation
        await asyncio.sleep(0.05)
        
        user_id = data.get('user_id')
        daily_total = 5500.0  # Mock: user has spent 5500 MYR today
        current_amount = float(data.get('amount', 0))
        
        projected_total = daily_total + current_amount
        
        risk_score = 0
        risk_factors = []
        
        if projected_total >= self.risk_thresholds['velocity_limit']:
            risk_score = 75
            risk_factors.append(f"Velocity limit exceeded: {projected_total:.2f} MYR/day")
        elif projected_total >= self.risk_thresholds['velocity_limit'] * 0.8:
            risk_score = 45
            risk_factors.append(f"High velocity: {projected_total:.2f} MYR/day")
        elif projected_total >= self.risk_thresholds['velocity_limit'] * 0.5:
            risk_score = 20
            risk_factors.append(f"Moderate velocity: {projected_total:.2f} MYR/day")
        else:
            risk_score = 5
            risk_factors.append(f"Normal velocity: {projected_total:.2f} MYR/day")
        
        return {
            'score': risk_score,
            'factors': risk_factors,
            'daily_total': daily_total,
            'projected_total': projected_total
        }
    
    async def _check_time_pattern_risk(self, data: Dict[str, Any]) -> Dict[str, Any]:
        """Check time-based risk patterns"""
        current_hour = datetime.utcnow().hour
        
        risk_score = 0
        risk_factors = []
        
        # High-risk hours (late night/early morning)
        if current_hour >= 23 or current_hour <= 5:
            risk_score = 30
            risk_factors.append(f"Late night transaction: {current_hour}:00")
        elif current_hour <= 7:
            risk_score = 15
            risk_factors.append(f"Early morning transaction: {current_hour}:00")
        else:
            risk_score = 5
            risk_factors.append(f"Normal hours transaction: {current_hour}:00")
        
        return {
            'score': risk_score,
            'factors': risk_factors,
            'transaction_hour': current_hour
        }
    
    async def _check_merchant_reputation_risk(self, data: Dict[str, Any]) -> Dict[str, Any]:
        """Check merchant reputation risk"""
        merchant_id = data.get('merchant_id')
        
        # Mock merchant reputation data
        merchant_reputation = {
            'MERCH_001': {'score': 95, 'category': 'trusted'},
            'MERCH_002': {'score': 60, 'category': 'moderate'},
            'UNKNOWN': {'score': 30, 'category': 'unknown'}
        }
        
        merchant_data = merchant_reputation.get(merchant_id, merchant_reputation['UNKNOWN'])
        merchant_score = merchant_data['score']
        
        # Invert score for risk (high reputation = low risk)
        risk_score = max(0, 100 - merchant_score) * 0.3
        
        risk_factors = [f"Merchant reputation: {merchant_data['category']} ({merchant_score}/100)"]
        
        return {
            'score': risk_score,
            'factors': risk_factors,
            'merchant_reputation': merchant_data
        }
    
    def _get_risk_level(self, risk_score: float) -> str:
        """Get risk level based on score"""
        if risk_score >= 85:
            return 'CRITICAL'
        elif risk_score >= 70:
            return 'HIGH'
        elif risk_score >= 40:
            return 'MEDIUM'
        elif risk_score >= 20:
            return 'LOW'
        else:
            return 'MINIMAL'
    
    def _determine_action(self, risk_score: float) -> Dict[str, Any]:
        """Determine action based on risk score"""
        if risk_score >= self.config['auto_block_threshold']:
            return {
                'action': 'BLOCK',
                'reason': 'High risk transaction blocked automatically',
                'require_approval': False
            }
        elif risk_score >= self.config['manual_review_threshold']:
            return {
                'action': 'MANUAL_REVIEW',
                'reason': 'Transaction requires manual review',
                'require_approval': True
            }
        else:
            return {
                'action': 'APPROVE',
                'reason': 'Low risk transaction approved',
                'require_approval': False
            }
    
    def validate_input(self, data: Dict[str, Any]) -> bool:
        """Validate input data for risk assessment"""
        required_fields = ['amount', 'user_id']
        
        for field in required_fields:
            if field not in data:
                return False
        
        # Validate amount
        try:
            amount = float(data['amount'])
            if amount <= 0:
                return False
        except (ValueError, TypeError):
            return False
        
        return True 
Risk Checker Plugin for SatuPay Payment Switch
Analyzes transaction risk and applies risk management rules
"""

import asyncio
from datetime import datetime, timedelta
from typing import Dict, Any, List
from .plugin_manager import BasePlugin

class RiskCheckerPlugin(BasePlugin):
    """Risk Checker Plugin for transaction risk assessment"""
    
    def __init__(self):
        super().__init__()
        self.name = "Risk Checker"
        self.version = "1.0.0"
        self.description = "Transaction risk assessment and fraud detection"
        self.critical = False
        
        # Risk thresholds
        self.risk_thresholds = {
            'high_amount': 10000.0,  # MYR
            'very_high_amount': 50000.0,  # MYR
            'suspicious_frequency': 10,  # transactions per hour
            'velocity_limit': 50000.0,  # MYR per day
            'international_limit': 25000.0,  # MYR for international
            'max_risk_score': 85  # Block if risk score > 85
        }
        
        # Risk factors weights
        self.risk_weights = {
            'amount_factor': 0.3,
            'frequency_factor': 0.25,
            'velocity_factor': 0.2,
            'time_factor': 0.1,
            'merchant_factor': 0.1,
            'location_factor': 0.05
        }
        
        self.config = {
            'enabled_checks': [
                'amount_check',
                'frequency_check', 
                'velocity_check',
                'time_pattern_check',
                'merchant_reputation_check'
            ],
            'auto_block_threshold': 85,
            'manual_review_threshold': 70,
            'risk_thresholds': self.risk_thresholds
        }
    
    async def execute(self, data: Dict[str, Any]) -> Dict[str, Any]:
        """Execute risk assessment"""
        try:
            risk_assessment = await self._perform_risk_assessment(data)
            
            # Determine action based on risk score
            action = self._determine_action(risk_assessment['risk_score'])
            
            return {
                'success': True,
                'data': {
                    'risk_assessment': risk_assessment,
                    'risk_action': action,
                    'risk_checks_completed': True
                }
            }
            
        except Exception as e:
            return {
                'success': False,
                'error': f"Risk assessment failed: {str(e)}",
                'critical': False
            }
    
    async def _perform_risk_assessment(self, data: Dict[str, Any]) -> Dict[str, Any]:
        """Perform comprehensive risk assessment"""
        
        risk_factors = {}
        total_risk_score = 0
        
        # Amount-based risk
        if 'amount_check' in self.config['enabled_checks']:
            amount_risk = await self._check_amount_risk(data)
            risk_factors['amount_risk'] = amount_risk
            total_risk_score += amount_risk['score'] * self.risk_weights['amount_factor']
        
        # Frequency-based risk
        if 'frequency_check' in self.config['enabled_checks']:
            frequency_risk = await self._check_frequency_risk(data)
            risk_factors['frequency_risk'] = frequency_risk
            total_risk_score += frequency_risk['score'] * self.risk_weights['frequency_factor']
        
        # Velocity-based risk
        if 'velocity_check' in self.config['enabled_checks']:
            velocity_risk = await self._check_velocity_risk(data)
            risk_factors['velocity_risk'] = velocity_risk
            total_risk_score += velocity_risk['score'] * self.risk_weights['velocity_factor']
        
        # Time pattern risk
        if 'time_pattern_check' in self.config['enabled_checks']:
            time_risk = await self._check_time_pattern_risk(data)
            risk_factors['time_risk'] = time_risk
            total_risk_score += time_risk['score'] * self.risk_weights['time_factor']
        
        # Merchant reputation risk
        if 'merchant_reputation_check' in self.config['enabled_checks']:
            merchant_risk = await self._check_merchant_reputation_risk(data)
            risk_factors['merchant_risk'] = merchant_risk
            total_risk_score += merchant_risk['score'] * self.risk_weights['merchant_factor']
        
        return {
            'risk_score': min(round(total_risk_score, 1), 100),
            'risk_level': self._get_risk_level(total_risk_score),
            'risk_factors': risk_factors,
            'assessment_timestamp': datetime.utcnow().isoformat()
        }
    
    async def _check_amount_risk(self, data: Dict[str, Any]) -> Dict[str, Any]:
        """Check amount-based risk"""
        amount = float(data.get('amount', 0))
        currency = data.get('currency', 'MYR')
        
        # Convert to MYR if needed (simplified)
        if currency != 'MYR':
            # Use rough conversion rates for risk assessment
            conversion_rates = {'USD': 4.7, 'SGD': 3.5, 'EUR': 5.1}
            amount = amount * conversion_rates.get(currency, 1)
        
        risk_score = 0
        risk_factors = []
        
        if amount >= self.risk_thresholds['very_high_amount']:
            risk_score = 80
            risk_factors.append(f"Very high amount: {amount:.2f} MYR")
        elif amount >= self.risk_thresholds['high_amount']:
            risk_score = 50
            risk_factors.append(f"High amount: {amount:.2f} MYR")
        elif amount >= self.risk_thresholds['high_amount'] * 0.5:
            risk_score = 25
            risk_factors.append(f"Moderate amount: {amount:.2f} MYR")
        else:
            risk_score = 5
            risk_factors.append(f"Normal amount: {amount:.2f} MYR")
        
        return {
            'score': risk_score,
            'factors': risk_factors,
            'amount_myr': amount
        }
    
    async def _check_frequency_risk(self, data: Dict[str, Any]) -> Dict[str, Any]:
        """Check frequency-based risk (mock implementation)"""
        # In production, this would query the database for recent transactions
        await asyncio.sleep(0.05)  # Simulate DB query
        
        # Mock frequency data
        user_id = data.get('user_id')
        recent_txn_count = 3  # Mock: user has 3 transactions in last hour
        
        risk_score = 0
        risk_factors = []
        
        if recent_txn_count >= self.risk_thresholds['suspicious_frequency']:
            risk_score = 70
            risk_factors.append(f"Suspicious frequency: {recent_txn_count} txns/hour")
        elif recent_txn_count >= self.risk_thresholds['suspicious_frequency'] * 0.6:
            risk_score = 40
            risk_factors.append(f"High frequency: {recent_txn_count} txns/hour")
        elif recent_txn_count >= 5:
            risk_score = 20
            risk_factors.append(f"Moderate frequency: {recent_txn_count} txns/hour")
        else:
            risk_score = 5
            risk_factors.append(f"Normal frequency: {recent_txn_count} txns/hour")
        
        return {
            'score': risk_score,
            'factors': risk_factors,
            'recent_transaction_count': recent_txn_count
        }
    
    async def _check_velocity_risk(self, data: Dict[str, Any]) -> Dict[str, Any]:
        """Check velocity-based risk (daily spending limit)"""
        # Mock implementation
        await asyncio.sleep(0.05)
        
        user_id = data.get('user_id')
        daily_total = 5500.0  # Mock: user has spent 5500 MYR today
        current_amount = float(data.get('amount', 0))
        
        projected_total = daily_total + current_amount
        
        risk_score = 0
        risk_factors = []
        
        if projected_total >= self.risk_thresholds['velocity_limit']:
            risk_score = 75
            risk_factors.append(f"Velocity limit exceeded: {projected_total:.2f} MYR/day")
        elif projected_total >= self.risk_thresholds['velocity_limit'] * 0.8:
            risk_score = 45
            risk_factors.append(f"High velocity: {projected_total:.2f} MYR/day")
        elif projected_total >= self.risk_thresholds['velocity_limit'] * 0.5:
            risk_score = 20
            risk_factors.append(f"Moderate velocity: {projected_total:.2f} MYR/day")
        else:
            risk_score = 5
            risk_factors.append(f"Normal velocity: {projected_total:.2f} MYR/day")
        
        return {
            'score': risk_score,
            'factors': risk_factors,
            'daily_total': daily_total,
            'projected_total': projected_total
        }
    
    async def _check_time_pattern_risk(self, data: Dict[str, Any]) -> Dict[str, Any]:
        """Check time-based risk patterns"""
        current_hour = datetime.utcnow().hour
        
        risk_score = 0
        risk_factors = []
        
        # High-risk hours (late night/early morning)
        if current_hour >= 23 or current_hour <= 5:
            risk_score = 30
            risk_factors.append(f"Late night transaction: {current_hour}:00")
        elif current_hour <= 7:
            risk_score = 15
            risk_factors.append(f"Early morning transaction: {current_hour}:00")
        else:
            risk_score = 5
            risk_factors.append(f"Normal hours transaction: {current_hour}:00")
        
        return {
            'score': risk_score,
            'factors': risk_factors,
            'transaction_hour': current_hour
        }
    
    async def _check_merchant_reputation_risk(self, data: Dict[str, Any]) -> Dict[str, Any]:
        """Check merchant reputation risk"""
        merchant_id = data.get('merchant_id')
        
        # Mock merchant reputation data
        merchant_reputation = {
            'MERCH_001': {'score': 95, 'category': 'trusted'},
            'MERCH_002': {'score': 60, 'category': 'moderate'},
            'UNKNOWN': {'score': 30, 'category': 'unknown'}
        }
        
        merchant_data = merchant_reputation.get(merchant_id, merchant_reputation['UNKNOWN'])
        merchant_score = merchant_data['score']
        
        # Invert score for risk (high reputation = low risk)
        risk_score = max(0, 100 - merchant_score) * 0.3
        
        risk_factors = [f"Merchant reputation: {merchant_data['category']} ({merchant_score}/100)"]
        
        return {
            'score': risk_score,
            'factors': risk_factors,
            'merchant_reputation': merchant_data
        }
    
    def _get_risk_level(self, risk_score: float) -> str:
        """Get risk level based on score"""
        if risk_score >= 85:
            return 'CRITICAL'
        elif risk_score >= 70:
            return 'HIGH'
        elif risk_score >= 40:
            return 'MEDIUM'
        elif risk_score >= 20:
            return 'LOW'
        else:
            return 'MINIMAL'
    
    def _determine_action(self, risk_score: float) -> Dict[str, Any]:
        """Determine action based on risk score"""
        if risk_score >= self.config['auto_block_threshold']:
            return {
                'action': 'BLOCK',
                'reason': 'High risk transaction blocked automatically',
                'require_approval': False
            }
        elif risk_score >= self.config['manual_review_threshold']:
            return {
                'action': 'MANUAL_REVIEW',
                'reason': 'Transaction requires manual review',
                'require_approval': True
            }
        else:
            return {
                'action': 'APPROVE',
                'reason': 'Low risk transaction approved',
                'require_approval': False
            }
    
    def validate_input(self, data: Dict[str, Any]) -> bool:
        """Validate input data for risk assessment"""
        required_fields = ['amount', 'user_id']
        
        for field in required_fields:
            if field not in data:
                return False
        
        # Validate amount
        try:
            amount = float(data['amount'])
            if amount <= 0:
                return False
        except (ValueError, TypeError):
            return False
        
        return True 